File: NowCoder/牛客小白月赛/牛客小白月赛106/普通DISCO_1.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, u, v;
    cin >> n;
    vector<vector<int>> g(n);
    for (int i = 0; i < n - 1; i++) {
        cin >> u >> v;
        u--; v--;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    vector<int> fa(n, -1);
    vector<int> dep(n, 1);
    auto dfs = [&](auto dfs, int u) -> void {
        for (int v : g[u]) {
            if (v == fa[u]) continue;
            fa[v] = u;
            dep[v] = dep[u] + 1;
            dfs(dfs, v);
        }
    };
    dfs(dfs, 0);

    int idx = 0;
    for (int u = 0; u < n; u++) {
        if (dep[u] > dep[idx]) idx = u;
    }

    queue<pair<int, int>> q;
    vector<bool> vis(n, false);
    int t = idx;
    while (t != -1) {
        q.push({0, t});
        vis[t] = true;
        t = fa[t];
    }

    int ans = 1;
    while (!q.empty()) {
        auto [d, u] = q.front(); q.pop();
        ans = max(ans, d);
        for (int v : g[u]) {
            if (vis[v]) continue;
            vis[v] = true;
            q.push({d + 1, v});
        }
    }
    cout << ans + dep[idx] - 1 << endl;
    return 0;
}

File: NowCoder/Contest/牛客小白月赛/牛客小白月赛106/普通DISCO_1.py
from collections import deque
 
n = int(input())
g = [[] for _ in range(n)]
for _ in range(n - 1):
    u, v = map(lambda x: int(x) - 1, input().split())
    g[u].append(v)
    g[v].append(u)
 
fa = [-1] * n
dep = [0] * n
q = deque([(0, -1)])
while q:
    u, f = q.popleft()
    fa[u] = f
    dep[u] = dep[f] + 1 if f != -1 else 1
    for v in g[u]:
        if v == f:
            continue
        q.append((v, u))
# def dfs(u):
#     for v in g[u]:
#         if v == fa[u]:
#             continue
#         fa[v] = u
#         dep[v] = dep[u] + 1
#         dfs(v)
# dfs(0)

idx = 0
for u in range(n):
    if dep[u] > dep[idx]:
        idx = u
 
q = deque()
vis = [False] * n
t = idx
while t != -1:
    q.append((0, t))
    vis[t] = True
    t = fa[t]
 
ans = 1 # 這裡至少要是 1
while q:
    d, u = q.popleft()
    ans = max(ans, d)
    for v in g[u]:
        if vis[v]:
            continue
        vis[v] = True
        q.append((d + 1, v))
 
print(ans + dep[idx] - 1)